### ISSUE TITLE
False report of TS2365

### ISSUE THREAD

**Issue by fuchen** [Microsoft/TypeScript#17881]
**TypeScript Version:**  2.4.2

**Code**

```ts
enum TestEnum {
  A, B, C
}

class TestClass {
  public foo: TestEnum

  public f1() {
    this.foo = TestEnum.A
    this.f2()
    if (this.foo === TestEnum.B) { // error TS2365: Operator '===' cannot be applied to types 'TestEnum.A' and 'TestEnum.B'.
      this.foo = TestEnum.C
    }
  }

  public f2() {
    this.foo = TestEnum.B
  }
}
```
**Actual behavior:**
tsc reports error TS2365.

**Expected behavior:**
Since `this.foo` might be changed in function `f2`, tsc should not report error TS2365.

**Comment by ahejlsberg** [comment]
See #9998.
